application/common/models/EmailQueue.php

```
@@ -0,0 +1,219 @@
+            $mailer = $this->getMailer();
+            $sent = $mailer->send();
+            if ($sent === 0 || $sent === false) {
+                throw new \Exception('Unable to send email', 1461011826);
+            }
+
+            /*
+             * Create event log entry if needed
+             */
+            $this->createEventLogEntry();
+
+            /*
+             * Remove entry from queue (if saved to queue) after successful send
+             */
+            $this->removeFromQueue();
+
+            /*
+             * Log success
+             */
+            $log['status'] = 'sent';
+            \Yii::warning($log, 'application');
+
+        } catch (\Exception $e) {
+            /*
+             * Send failed, attempt to queue
+             */
+            $this->attempts_count += 1;
+            $this->last_attempt = Utils::getDatetime();
+            $this->queue($log, $e);
+        }
+    }
+
+    /**
+     * Builds a mailer object from $this and returns it
+     * @return \yii\mail\MessageInterface
+     */
+    private function getMailer()
+    {
+        $mailer = \Yii::$app->mailer->compose();
+        $mailer->setFrom(\Yii::$app->params['fromEmail']);
+        $mailer->setTo($this->to_address);
+        $mailer->setSubject($this->subject);
+        $mailer->setTextBody($this->text_body);
+
+        /*
+         * Conditionally set optional fields
+         */
+        $setMethods = [
+            'setCc' => $this->cc_address,
+            'setHtmlBody' => $this->html_body,
+        ];
+        foreach ($setMethods as $method => $value) {
+            if ($value) {
+                $mailer->$method($value);
+            }
+        }
+        
+        return $mailer;
+    }
+
+    /**
+     * Creates an EventLog entry if $this->event_log_* properties are set
+     */
+    private function createEventLogEntry()
+    {
+        if ($this->event_log_topic !== null && $this->event_log_details !== null && $this->event_log_user_id !== null) {
```

Should there be some indication to the code calling this if no event log entry ends up being created? I'm not sure, it just seems like a potential for silent failure.